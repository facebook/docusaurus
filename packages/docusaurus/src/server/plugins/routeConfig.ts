/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import {
  applyTrailingSlash,
  type ApplyTrailingSlashParams,
} from '@docusaurus/utils-common';
import type {RouteConfig} from '@docusaurus/types';

/** Recursively applies trailing slash config to all nested routes. */
export function applyRouteTrailingSlash<Route extends RouteConfig>(
  route: Route,
  params: ApplyTrailingSlashParams,
): Route {
  return {
    ...route,
    path: applyTrailingSlash(route.path, params),
    ...(route.routes && {
      routes: route.routes.map((subroute) =>
        applyRouteTrailingSlash(subroute, params),
      ),
    }),
  };
}

export function sortRoutes<Route extends RouteConfig>(
  routesToSort: Route[],
  baseUrl: string,
): Route[] {
  const routeConfigs = [...routesToSort];
  // Sort the route config. This ensures that route with nested
  // routes is always placed last.
  routeConfigs.sort((a, b) => {
    // Root route should get placed last.
    if (a.path === baseUrl && b.path !== baseUrl) {
      return 1;
    }
    if (a.path !== baseUrl && b.path === baseUrl) {
      return -1;
    }

    if (a.routes && !b.routes) {
      return 1;
    }
    if (!a.routes && b.routes) {
      return -1;
    }

    // If both are parent routes (for example routeBasePath: "/" and "/docs/"
    // We must order them carefully in case of overlapping paths
    if (a.routes && b.routes) {
      if (a.path === b.path) {
        // We don't really support that kind of routing ATM
        // React-Router by default will only "enter" a single parent route
      } else {
        if (a.path.includes(b.path)) {
          return -1;
        }
        if (b.path.includes(a.path)) {
          return 1;
        }
      }
    }

    // Higher priority get placed first.
    if (a.priority || b.priority) {
      const priorityA = a.priority ?? 0;
      const priorityB = b.priority ?? 0;
      const score = priorityB - priorityA;

      if (score !== 0) {
        return score;
      }
    }

    return a.path.localeCompare(b.path);
  });

  routeConfigs.forEach((routeConfig) => {
    if (routeConfig.routes) {
      routeConfig.routes = sortRoutes(routeConfig.routes, baseUrl);
    }
  });

  return routeConfigs;
}
