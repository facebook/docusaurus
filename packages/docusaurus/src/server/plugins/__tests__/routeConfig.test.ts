/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import {applyRouteTrailingSlash, sortRoutes} from '../routeConfig';
import type {RouteConfig} from '@docusaurus/types';
import type {ApplyTrailingSlashParams} from '@docusaurus/utils-common';

function route(path: string, subRoutes?: string[]): RouteConfig {
  const result: RouteConfig = {path, component: 'any'};

  if (subRoutes) {
    result.routes = subRoutes.map((subRoute) => route(subRoute));
  }

  return result;
}

function params(
  trailingSlash: boolean | undefined,
  baseUrl: string = '/',
): ApplyTrailingSlashParams {
  return {trailingSlash, baseUrl};
}

describe('applyRouteTrailingSlash', () => {
  it('apply to empty', () => {
    expect(applyRouteTrailingSlash(route(''), params(true))).toEqual(
      route('/'),
    );
    expect(applyRouteTrailingSlash(route(''), params(false))).toEqual(
      route(''),
    );
    expect(applyRouteTrailingSlash(route(''), params(undefined))).toEqual(
      route(''),
    );
  });

  it('apply to /', () => {
    expect(applyRouteTrailingSlash(route('/'), params(true))).toEqual(
      route('/'),
    );
    expect(applyRouteTrailingSlash(route('/'), params(false))).toEqual(
      route('/'),
    );
    expect(applyRouteTrailingSlash(route('/'), params(undefined))).toEqual(
      route('/'),
    );
  });

  it('apply to /abc', () => {
    expect(applyRouteTrailingSlash(route('/abc'), params(true))).toEqual(
      route('/abc/'),
    );
    expect(applyRouteTrailingSlash(route('/abc'), params(false))).toEqual(
      route('/abc'),
    );
    expect(applyRouteTrailingSlash(route('/abc'), params(undefined))).toEqual(
      route('/abc'),
    );
  });

  it('apply to /abc/', () => {
    expect(applyRouteTrailingSlash(route('/abc/'), params(true))).toEqual(
      route('/abc/'),
    );
    expect(applyRouteTrailingSlash(route('/abc/'), params(false))).toEqual(
      route('/abc'),
    );
    expect(applyRouteTrailingSlash(route('/abc/'), params(undefined))).toEqual(
      route('/abc/'),
    );
  });

  it('apply to /abc?search#anchor', () => {
    expect(
      applyRouteTrailingSlash(route('/abc?search#anchor'), params(true)),
    ).toEqual(route('/abc/?search#anchor'));
    expect(
      applyRouteTrailingSlash(route('/abc?search#anchor'), params(false)),
    ).toEqual(route('/abc?search#anchor'));
    expect(
      applyRouteTrailingSlash(route('/abc?search#anchor'), params(undefined)),
    ).toEqual(route('/abc?search#anchor'));
  });

  it('apply to /abc/?search#anchor', () => {
    expect(
      applyRouteTrailingSlash(route('/abc/?search#anchor'), params(true)),
    ).toEqual(route('/abc/?search#anchor'));
    expect(
      applyRouteTrailingSlash(route('/abc/?search#anchor'), params(false)),
    ).toEqual(route('/abc?search#anchor'));
    expect(
      applyRouteTrailingSlash(route('/abc/?search#anchor'), params(undefined)),
    ).toEqual(route('/abc/?search#anchor'));
  });

  it('not apply to /abc/?search#anchor when baseUrl=/abc/', () => {
    const baseUrl = '/abc/';
    expect(
      applyRouteTrailingSlash(
        route('/abc/?search#anchor'),
        params(true, baseUrl),
      ),
    ).toEqual(route('/abc/?search#anchor'));
    expect(
      applyRouteTrailingSlash(
        route('/abc/?search#anchor'),
        params(false, baseUrl),
      ),
    ).toEqual(route('/abc/?search#anchor'));
    expect(
      applyRouteTrailingSlash(
        route('/abc/?search#anchor'),
        params(undefined, baseUrl),
      ),
    ).toEqual(route('/abc/?search#anchor'));
  });

  it('apply to subroutes', () => {
    expect(
      applyRouteTrailingSlash(
        route('/abc', ['/abc/1', '/abc/2']),
        params(true),
      ),
    ).toEqual(route('/abc/', ['/abc/1/', '/abc/2/']));
    expect(
      applyRouteTrailingSlash(
        route('/abc', ['/abc/1', '/abc/2']),
        params(false),
      ),
    ).toEqual(route('/abc', ['/abc/1', '/abc/2']));
    expect(
      applyRouteTrailingSlash(
        route('/abc', ['/abc/1', '/abc/2']),
        params(undefined),
      ),
    ).toEqual(route('/abc', ['/abc/1', '/abc/2']));
  });

  it('apply for complex case', () => {
    expect(
      applyRouteTrailingSlash(
        route('/abc?search#anchor', ['/abc/1?search', '/abc/2#anchor']),
        params(true),
      ),
    ).toEqual(
      route('/abc/?search#anchor', ['/abc/1/?search', '/abc/2/#anchor']),
    );
  });

  it('apply for complex case with baseUrl', () => {
    const baseUrl = '/abc/';
    expect(
      applyRouteTrailingSlash(
        route('/abc/?search#anchor', ['/abc/1?search', '/abc/2#anchor']),
        params(false, baseUrl),
      ),
    ).toEqual(route('/abc/?search#anchor', ['/abc/1?search', '/abc/2#anchor']));
  });
});

describe('sortRoutes', () => {
  it('sorts route config correctly', () => {
    const routes: RouteConfig[] = [
      {
        path: '/',
        component: '',
        routes: [
          {path: '/someDoc', component: ''},
          {path: '/someOtherDoc', component: ''},
        ],
      },
      {
        path: '/',
        component: '',
      },
      {
        path: '/',
        component: '',
        routes: [{path: '/subroute', component: ''}],
      },
      {
        path: '/docs',
        component: '',
        routes: [
          {path: '/docs/someDoc', component: ''},
          {path: '/docs/someOtherDoc', component: ''},
        ],
      },
      {
        path: '/community',
        component: '',
      },
      {
        path: '/some-page',
        component: '',
      },
    ];

    expect(sortRoutes(routes, '/')).toMatchSnapshot();
  });

  it('sorts route config recursively', () => {
    const routes: RouteConfig[] = [
      {
        path: '/docs',
        component: '',
        routes: [
          {
            path: '/docs/tags',
            component: '',
            exact: true,
          },
          {
            path: '/docs',
            component: '',
            routes: [
              {
                path: '/docs/doc1',
                component: '',
                exact: true,
              },
              {
                path: '/docs/doc2',
                component: '',
                exact: true,
              },
            ],
          },
          {
            path: '/docs/tags/someTag',
            component: '',
            exact: true,
          },
        ],
      },
      {
        path: '/some/page',
        component: '',
        exact: true,
      },
    ];

    expect(sortRoutes(routes, '/')).toMatchSnapshot();
  });

  it('sorts route config given a baseURL', () => {
    const baseURL = '/latest/';
    const routes: RouteConfig[] = [
      {
        path: baseURL,
        component: '',
        routes: [
          {path: `${baseURL}someDoc`, component: ''},
          {path: `${baseURL}someOtherDoc`, component: ''},
        ],
      },
      {
        path: `${baseURL}example`,
        component: '',
      },
      {
        path: `${baseURL}docs`,
        component: '',
        routes: [
          {path: `${baseURL}docs/someDoc`, component: ''},
          {path: `${baseURL}docs/someOtherDoc`, component: ''},
        ],
      },
      {
        path: `${baseURL}community`,
        component: '',
      },
      {
        path: `${baseURL}some-page`,
        component: '',
      },
      {
        path: `${baseURL}`,
        component: '',
      },
    ];

    expect(sortRoutes(routes, baseURL)).toMatchSnapshot();
  });

  it('sorts parent route configs when one included in another', () => {
    const r1: RouteConfig = {
      path: '/one',
      component: '',
      routes: [{path: `/one/myDoc`, component: ''}],
    };
    const r2: RouteConfig = {
      path: '/',
      component: '',
      routes: [{path: `/someDoc`, component: ''}],
    };
    const r3: RouteConfig = {
      path: '/one/another',
      component: '',
      routes: [{path: `/one/another/myDoc`, component: ''}],
    };

    expect(sortRoutes([r1, r2, r3], '/')).toEqual([r3, r1, r2]);
    expect(sortRoutes([r3, r1, r2], '/')).toEqual([r3, r1, r2]);
  });
});
