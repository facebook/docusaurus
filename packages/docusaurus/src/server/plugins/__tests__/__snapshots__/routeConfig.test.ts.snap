// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sortRoutes sorts route config correctly 1`] = `
[
  {
    "component": "",
    "path": "/community",
  },
  {
    "component": "",
    "path": "/some-page",
  },
  {
    "component": "",
    "path": "/docs",
    "routes": [
      {
        "component": "",
        "path": "/docs/someDoc",
      },
      {
        "component": "",
        "path": "/docs/someOtherDoc",
      },
    ],
  },
  {
    "component": "",
    "path": "/",
  },
  {
    "component": "",
    "path": "/",
    "routes": [
      {
        "component": "",
        "path": "/someDoc",
      },
      {
        "component": "",
        "path": "/someOtherDoc",
      },
    ],
  },
  {
    "component": "",
    "path": "/",
    "routes": [
      {
        "component": "",
        "path": "/subroute",
      },
    ],
  },
]
`;

exports[`sortRoutes sorts route config given a baseURL 1`] = `
[
  {
    "component": "",
    "path": "/latest/community",
  },
  {
    "component": "",
    "path": "/latest/example",
  },
  {
    "component": "",
    "path": "/latest/some-page",
  },
  {
    "component": "",
    "path": "/latest/docs",
    "routes": [
      {
        "component": "",
        "path": "/latest/docs/someDoc",
      },
      {
        "component": "",
        "path": "/latest/docs/someOtherDoc",
      },
    ],
  },
  {
    "component": "",
    "path": "/latest/",
  },
  {
    "component": "",
    "path": "/latest/",
    "routes": [
      {
        "component": "",
        "path": "/latest/someDoc",
      },
      {
        "component": "",
        "path": "/latest/someOtherDoc",
      },
    ],
  },
]
`;

exports[`sortRoutes sorts route config recursively 1`] = `
[
  {
    "component": "",
    "exact": true,
    "path": "/some/page",
  },
  {
    "component": "",
    "path": "/docs",
    "routes": [
      {
        "component": "",
        "exact": true,
        "path": "/docs/tags",
      },
      {
        "component": "",
        "exact": true,
        "path": "/docs/tags/someTag",
      },
      {
        "component": "",
        "path": "/docs",
        "routes": [
          {
            "component": "",
            "exact": true,
            "path": "/docs/doc1",
          },
          {
            "component": "",
            "exact": true,
            "path": "/docs/doc2",
          },
        ],
      },
    ],
  },
]
`;

exports[`sortRoutes sorts route config correctly 1`] = `
[
  {
    "component": "",
    "path": "/community",
  },
  {
    "component": "",
    "path": "/some-page",
  },
  {
    "component": "",
    "path": "/docs",
    "routes": [
      {
        "component": "",
        "path": "/docs/someDoc",
      },
      {
        "component": "",
        "path": "/docs/someOtherDoc",
      },
    ],
  },
  {
    "component": "",
    "path": "/",
  },
  {
    "component": "",
    "path": "/",
    "routes": [
      {
        "component": "",
        "path": "/someDoc",
      },
      {
        "component": "",
        "path": "/someOtherDoc",
      },
    ],
  },
  {
    "component": "",
    "path": "/",
    "routes": [
      {
        "component": "",
        "path": "/subroute",
      },
    ],
  },
]
`;

exports[`sortRoutes sorts route config given a baseURL 1`] = `
[
  {
    "component": "",
    "path": "/latest/community",
  },
  {
    "component": "",
    "path": "/latest/example",
  },
  {
    "component": "",
    "path": "/latest/some-page",
  },
  {
    "component": "",
    "path": "/latest/docs",
    "routes": [
      {
        "component": "",
        "path": "/latest/docs/someDoc",
      },
      {
        "component": "",
        "path": "/latest/docs/someOtherDoc",
      },
    ],
  },
  {
    "component": "",
    "path": "/latest/",
  },
  {
    "component": "",
    "path": "/latest/",
    "routes": [
      {
        "component": "",
        "path": "/latest/someDoc",
      },
      {
        "component": "",
        "path": "/latest/someOtherDoc",
      },
    ],
  },
]
`;

exports[`sortRoutes sorts route config recursively 1`] = `
[
  {
    "component": "",
    "exact": true,
    "path": "/some/page",
  },
  {
    "component": "",
    "path": "/docs",
    "routes": [
      {
        "component": "",
        "exact": true,
        "path": "/docs/tags",
      },
      {
        "component": "",
        "exact": true,
        "path": "/docs/tags/someTag",
      },
      {
        "component": "",
        "path": "/docs",
        "routes": [
          {
            "component": "",
            "exact": true,
            "path": "/docs/doc1",
          },
          {
            "component": "",
            "exact": true,
            "path": "/docs/doc2",
          },
        ],
      },
    ],
  },
]
`;
